omod/src/test/java/org/openmrs/module/aijar/htmlformentry/DSDMProgramSubmissionActionTest.java

```diff
@@ -0,0 +1,254 @@
+
+    private String xml2 = "<htmlform>\n" + "" + "Date: <encounterDate default='2015-02-10 00:00:00'/>\n" + "Location: <encounterLocation default='1'/>\n" + "Provider: <encounterProvider role='Provider' default='2'/>\n" + "<input type=\"hidden\" name=\"personId\" value=\"1393\"/>" + "<input type=\"hidden\" name=\"createVisit\" value=\"false\"/>\n" + "<input type=\"hidden\" name=\"encounterId\" value=\"31182\"/>\n" + "<input type=\"hidden\" name=\"visitId\" value=\"31165\"/>\n" + "<obs id=\"165143\" conceptId=\"165143\" answerConceptIds=\"165138,165140,165139,165142,165141\" answerLabels=\"FBIM,FBG,FTR,CDDP,CCLAD\" required=\"required\" labelText=\"DSDM Model\" defaultValue=\"165140\"/>\n" + "<postSubmissionAction class='org.openmrs.module.aijar.htmlformentry.DSDSProgramSubmissionAction'/>\n" + "<submit/>" + "</htmlform>";
+
+    @Before
+    public void setup() throws Exception {
+        executeDataSet(UGANDAEMR_DSDM_DATASET_XML);
+    }
+
+    @After
+    public void cleanup() throws Exception {
+        deleteAllData();
+    }
+
+    /**
+     * New  Patient Encounter with same DSDM Model as the current Model
+     *
+     * @throws Exception
+     */
+    @Test
+    public void shouldEnrollPatientInDSDMProgramWhenDSDMIsSelectedOnNewForm() throws Exception {
+        Patient patient = new Patient(1393);
+        ProgramWorkflowService service = Context.getService(ProgramWorkflowService.class);
+
+        //prepare and submit an html form to enroll patient in mch program
+        HtmlForm htmlForm1 = new HtmlForm();
+        htmlForm1.setXmlData(xml);
+        Form form1 = new Form(8);
+        form1.setEncounterType(new EncounterType(15));
+        htmlForm1.setForm(form1);
+        FormEntrySession session1 = new FormEntrySession(patient, null, FormEntryContext.Mode.ENTER, htmlForm1, new MockHttpSession());
+
+
+        //getHtmlToDisplay() is called to generate necessary tag handlers and cache the form
+        session1.getHtmlToDisplay();
+
+        //prepareForSubmit is called to set patient and encounter if specified in tags
+        session1.prepareForSubmit();
+
+        HttpServletRequest request = mock(MockHttpServletRequest.class);
+        when(request.getParameter("w1")).thenReturn("2017-04-01");
+        //Seleted DSDM Program in form is FTR
+        when(request.getParameter("w8")).thenReturn("165139");
+        session1.getSubmissionController().handleFormSubmission(session1, request);
+
+        session1.applyActions();
+        List<PatientProgram> patientProgramList = getPatientPrograms(patient, null,null,null,false);
+        Assert.assertEquals(service.getProgramByName("FTR"), patientProgramList.get(1).getProgram());
+        Assert.assertEquals(service.getProgramByName("FBIM"), patientProgramList.get(0).getProgram());
+        Assert.assertEquals(patientProgramList.get(1).getDateEnrolled(), patientProgramList.get(0).getDateCompleted());
```

Comment:
@slubwama Break into 2 Assertions using the hardcoded date in line 72 - that way you are sure that the dates are correct as both could be null and this test still passes

Reply:
@ssmusoke that is done